### PR TITLE
fix: middle route recipient

### PR DESCRIPTION
**Pull request by ChefJerry**
### Summary
📝♻️🛠️

This pull request improves the code for mixed swap commands in the `pancakeswap.ts` file. It introduces a constant `CONTRACT_BALANCE` in the `constants.ts` file to simplify the input amount logic. It also makes the code more readable and maintainable by using constants, variables, and a helper function.

> _We're sailing on the smart router, aye, aye, aye_
> _We're swapping tokens with `CONTRACT_BALANCE`_
> _We've refactored all the code, it's neat and clear_
> _We've used constants, variables, and a helper function here_

### Walkthrough
*  Define a constant `CONTRACT_BALANCE` to represent the maximum balance of the smart router contract ([link](https://github.com/pancakeswap/pancake-frontend/pull/8112/files?diff=unified&w=0#diff-e86aea0e9497b4caca196310dc3733ea2a0792917a7e238ac9e0091009278a20R132))
* Import `CONTRACT_BALANCE` in `pancakeswap.ts` and use it to indicate that the contract should use its own balance as the input amount for a swap section ([link](https://github.com/pancakeswap/pancake-frontend/pull/8112/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L13-R13))
* Refactor the `addMixedSwap` function in `pancakeswap.ts` to use a switch statement based on the route type, a local variable `isLastSection` to check if the section is the last one in the route, and an array `amounts` to pass the input and output amounts to the commands ([link](https://github.com/pancakeswap/pancake-frontend/pull/8112/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L302-L305), [link](https://github.com/pancakeswap/pancake-frontend/pull/8112/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05L316-R371))
* Extract the logic for choosing the recipient address for each swap section into a helper function `getRecipient` and a local variable `payByUser` in `pancakeswap.ts` ([link](https://github.com/pancakeswap/pancake-frontend/pull/8112/files?diff=unified&w=0#diff-566cd8974b99b34b5fdc1c4a109c3da8754c475cba38788149605d459e291b05R307-R322))

### CODE REVIEW

**Review by chefjackson — APPROVED** [review]

File: packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts [hunk 1]
```diff
@@ -299,89 +299,76 @@ async function addMixedSwap(
...
       let outputToken
       let inputToken = inputAmount.currency.wrapped
 
       for (let i = 0; i < sections.length; i++) {
         const section = sections[i]
+        const nextSection = sections[i + 1] ?? []
+        const isFirstSection = i == 0
+        const isLastSection = i === sections.length - 1
+
+        const nextIsV2 = nextSection.every(SmartRouter.isV2Pool)
```
Comment (chefjackson) [inline packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:311]: `nextIsV2` will be true if `nextSection` is an empty array?

Comment (ChefJerry) [reply packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:311]: nope, 
```ts
export function isV2Pool(pool: Pool): pool is V2Pool {
  return pool.type === PoolType.V2
}
```

Comment (ChefJerry) [reply packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts:311]: em... wait
image

**Review by chefjackson — CHANGES_REQUESTED** [review]
kindly check the comment

**Review by chefjackson — APPROVED** [review]

**Pull request merged**